### PR TITLE
Original vs hedged request metrics

### LINKED ISSUE #42: Determining the effectiveness of hedging
Hey @cristaloleg

Is there a way to determine how many hedged requests actually "won" vs the original request, with the stats?

### PR DESCRIPTION
**Pull request by cristaloleg**
Fixes #42

### CODE REVIEW

**Review by dannykopping — APPROVED** [review]
Exactly how I was planning to implement it 🙂 
LGTM!

**Comment by cristaloleg** [comment]
Thanks, then I will add tests to cover this case and release a new version soon.

**Comment by dannykopping** [comment]
Really appreciate it @cristaloleg 🙌

File: hedged.go [hunk 1]
```diff
@@ -140,6 +140,11 @@ func (ht *hedgedTransport) RoundTrip(req *http.Request) (*http.Response, error)
 		switch {
 		case resp.Resp != nil:
 			resultIdx = resp.Index
+			if resultIdx == 0 {
+				ht.metrics.originalRequestWinsInc()
```
Comment (joe-elliott) [inline hedged.go:144]: is there a reason to record both of these? isn't one deriveable from the other?

total requests - original requests wins = hedged requests wins

Comment (cristaloleg) [reply hedged.go:144]: Well, yeah, just with the extra steps.

I don't have a strong opinion on that, the internal `ht.metrics` can have only 1 atomic counter but there can be 2 exposed metrics.

@dannykopping @storozhukBM wdyt?

Comment (dannykopping) [reply hedged.go:144]: I'm not sure.
For example:

=> request
  => original... slow
  => hedge 1... slow
  => hedge 2.. fast => wins!

That would be 3 total requests - 0 original request wins = 3 hedged request wins...?

Comment (cristaloleg) [reply hedged.go:144]: WOW, good catch.

**Comment by cristaloleg** [comment]
Tests are added @dannykopping @joe-elliott . Not sure about thread above, IMO having 2 counters is much simpler and costs nothing, WDYT ?

I will refactor tests in the next PR, looks too wordy now.

File: hedged_test.go [hunk 2]
```diff
@@ -257,6 +257,128 @@ func TestBestResponse(t *testing.T) {
...
+func TestOriginalResponseWins(t *testing.T) {
+	const shortest = 20 * time.Millisecond
+	timeouts := [...]time.Duration{shortest, 30 * shortest, 5 * shortest, shortest, shortest, shortest}
+	timeoutCh := make(chan time.Duration, len(timeouts))
+	for _, t := range timeouts {
+		timeoutCh <- t
+	}
+
+	url := testServerURL(t, func(w http.ResponseWriter, r *http.Request) {
+		time.Sleep(<-timeoutCh)
```
Comment (dannykopping) [inline hedged_test.go:269]: Oh this is clever

File: hedged_test.go [hunk 3]
```diff
@@ -257,6 +257,128 @@ func TestBestResponse(t *testing.T) {
...
+		t.Fatal(err)
+	}
+	defer resp.Body.Close()
+
+	passed := time.Since(start)
+
+	if float64(passed) > float64(shortest)*2.5 {
+		t.Fatalf("want %v, got %v", shortest, passed)
+	}
+	if requestedRoundTrips := metrics.RequestedRoundTrips(); requestedRoundTrips != 1 {
```
Comment (dannykopping) [inline hedged_test.go:298]: Nit: if we had a way to print all stats, a string comparison might be more readable

File: examples_test.go [hunk 4]
```diff
@@ -57,7 +58,7 @@ func ExampleRoundTripper() {
 	go func() {
 		for {
 			time.Sleep(time.Second)
-			fmt.Printf("stats: %+v\n", stats)
+			fmt.Fprintf(io.Discard, "all requests: %d\n", stats.ActualRoundTrips())
```
Comment (cristaloleg) [inline examples_test.go:61]: Ok to ignore, still just an example.

**Comment by dannykopping** [comment]
🎉🎉🎉

**Comment by dannykopping** [comment]
I tried to integrate this, but long story short we have many hedged clients but only register the metrics once; this makes it difficult to use the stats because one client might have `0` for `HedgedRequestWins()` and another might have some other value - which one would we use to update the metrics? The answer is: without major refactoring, we can't.

I solved this by using the return value of the roundtrip:


```go
func (rt *limitedHedgingRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
	isHedged := hedgedhttp.IsHedgedRequest(req)
	if isHedged {
		if !rt.limiter.Allow() {
			totalRateLimitedHedgeRequests.Inc()
			return nil, ErrTooManyHedgeRequests
		}
		totalHedgeRequests.Inc()
	}

	resp, err := rt.next.RoundTrip(req)
	if err == nil {
		if isHedged {
			requestsWon.WithLabelValues("hedged").Inc()
		} else {
			requestsWon.WithLabelValues("original").Inc()
		}
	}

	return resp, err
}
```

**Comment by cristaloleg** [comment]
Thanks for the update! So...as I understood it's not library's problem but only a case that you've many hedged clients in 1 app. Am I right?

BTW, any numbers how many wons? :D (such info is probably under NDA but I decided to try :D )

**Comment by dannykopping** [comment]
> So...as I understood it's not library's problem but only a case that you've many hedged clients in 1 app. Am I right?

That's correct!

> BTW, any numbers how many wons? :D (such info is probably under NDA but I decided to try :D )

I'll update you once I roll this out to production next week 🙂 I'm sure I can send a % of hedging effectiveness with no problems.

**Comment by dannykopping** [comment]
image
image

This is from a 12 hour range in our pre-prod environment. Seems like we're averaging about a 20% effectiveness rate (i.e. only 1 of every 5 hedged requests wins vs the original).

Do you have a good intuition about what this rate _should_ be in order to justify the extra expense?

**Comment by cristaloleg** [comment]
> Do you have a good intuition about what this rate should be in order to justify the extra expense?

Sorry, no real formulas, someone with a better probability theory background should comment on that.

But the intuition suggests that it should not be high (or even lower than that). Request hedging is only about tail latency, so basically 1-5% of the requests.

I don't think there is are exact numbers for everyone. Sleep between calls, amount of hedged calls and success rate as a result heavily depends on the systems and it's behaviour. The only thing I can really suggest is to play with the numbers.

Probably the target latency should be a const (so, SLO) and after that tweaking sleeps & amount should be adjusted to minimise numbers of calls but keeping latency at the desired level. My 2c.

CC: @storozhukBM
